source/endpoints/RuleEngine.cpp

```diff
@@ -0,0 +1,177 @@
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/common/string.h>
+#include <aws/crt/Api.h>
+#include <aws/crt/endpoints/RuleEngine.h>
+#include <aws/sdkutils/endpoints_rule_engine.h>
+#include <aws/sdkutils/partitions.h>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        namespace Endpoints
+        {
+            RequestContext::RequestContext(Allocator *allocator) noexcept : m_allocator(allocator)
+            {
+                m_requestContext = aws_endpoints_request_context_new(allocator);
+            }
+
+            RequestContext::~RequestContext()
+            {
+                m_requestContext = aws_endpoints_request_context_release(m_requestContext);
+            }
+
+            bool RequestContext::AddString(const ByteCursor &name, const ByteCursor &value)
+            {
+                return AWS_OP_SUCCESS !=
+                       aws_endpoints_request_context_add_string(m_allocator, m_requestContext, name, value);
+            }
+
+            bool RequestContext::AddBoolean(const ByteCursor &name, bool value)
+            {
+                return AWS_OP_SUCCESS !=
+                       aws_endpoints_request_context_add_boolean(m_allocator, m_requestContext, name, value);
+            }
+
+            ResolutionOutcome::ResolutionOutcome(aws_endpoints_resolved_endpoint *impl, Allocator *allocator)
+                : m_allocator(allocator), m_resolvedEndpoint(impl)
+            {
+            }
+
+            ResolutionOutcome::ResolutionOutcome(ResolutionOutcome &&toMove) noexcept
+                : m_resolvedEndpoint(toMove.m_resolvedEndpoint)
+            {
+                toMove.m_resolvedEndpoint = nullptr;
+            }
+
+            ResolutionOutcome &ResolutionOutcome::operator=(ResolutionOutcome &&toMove)
+            {
+                if (&toMove != this)
+                {
+                    *this = ResolutionOutcome(std::move(toMove));
+                }
+
+                return *this;
+            }
+
+            ResolutionOutcome::~ResolutionOutcome()
+            {
+                aws_endpoints_resolved_endpoint_release(m_resolvedEndpoint);
+            }
+
+            bool ResolutionOutcome::IsEndpoint() const noexcept
+            {
+                return AWS_ENDPOINTS_RESOLVED_ENDPOINT == aws_endpoints_resolved_endpoint_get_type(m_resolvedEndpoint);
+            }
+
+            bool ResolutionOutcome::IsError() const noexcept
+            {
+                return AWS_ENDPOINTS_RESOLVED_ERROR == aws_endpoints_resolved_endpoint_get_type(m_resolvedEndpoint);
+            }
+
+            Optional<StringView> ResolutionOutcome::GetUrl() const noexcept
+            {
+                ByteCursor url;
+                if (aws_endpoints_resolved_endpoint_get_url(m_resolvedEndpoint, &url))
+                {
+                    return Optional<StringView>();
+                }
+
+                return Optional<StringView>(ByteCursorToStringView(url));
```

Comment:
both c-tors with arguments of Crt::Optional
```
Optional(const T &val)
Optional(T &&val)
```
have a call to `new`
and `new` can throw at least std::bad_alloc

Reply:
hmm, good point. i see we are doing noexcept in other cases where we create optional (ex. http classes). will remove noexcept